### PR TITLE
[WIP] Fix #2090: Endpoint to get remaining submissions limits for all phases and Submission Page for Docker Based Challenges

### LINKED ISSUE #2090: Disable submissions from UI if the challenge is Docker based
## Current Scenario

Currently, a user can submit to both docker based and non-docker based challenges using the UI. We'd like to restrict the user to submit to docker based challenges using the CLI.

## Expected Behaviour

The user can't submit to a challenge using the UI.

## Deliverables

- [ ] Add a check in the `challenge_submission` API to check if the challenge is docker based or not. If yes, then don't create a submission instead return a response saying `The challenge {} requires uploading code in the form of docker images. Please use EvalAI-CLI to make a submission.format(challenge.title)`. The response should be returned for both challenge host as well as participants.

- [ ] Add the test cases
- [ ] Add a check on UI using the field `is_docker_based` from the challenge model to disable all the form attributes in the submission page if `is_docker_based == true`.

Please feel free to ask in case of doubts.

### PR DESCRIPTION
**Pull request by vkartik97**
Fixes: #2090 

**Deliverables**: 
- [x] Added a check in the `challenge_submission` API for `is_docker_based` for Docker based Challenge submissions.
- [x]  ~Added a check on UI using the field `is_docker_based` from the challenge model to hide the form and display `The challenge *title* requires uploading code in the form of docker images. Please use EvalAI-CLI to make a submission` below `Submission Guidelines`  in the submission page if `is_docker_based == true`.~
- [x]  Modified Tests

**Additions**:
- [x] Added `Instructions to Submit Docker Image using EvalAI-CLI` for Docker Submissions
- [x] Made Endpoint to get Remaining Submissions Limits for all *phases*.
- [x] Table to Display Remaining Submission Limits for all Phases for  Docker Based Challenge
- [x] Show/Hide Auth Code for `evalai get_token` on  `Instructions to Submit Docker Image using EvalAI-CLI`
- [x] Added Tests

**API**
`GET jobs/*challenge_pk*/remaining_submissions`
```json
{
    "participant_team": "Sample_Participant_Team",
    "participant_team_id": 2,
    "phases": [
        {
            "id": 1,
            "name": "Megan Phase",
            "start_date": "2018-10-28T14:22:53.022639Z",
            "end_date": "2020-06-19T14:22:53.022660Z",
            "message": {
                "remaining_submissions_this_month_count": 9,
                "remaining_submissions_today_count": 5,
                "remaining_submissions": 29
            }
        },
        {
            "id": 2,
            "name": "Molly Phase",
            "start_date": "2018-10-28T14:22:53Z",
            "end_date": "2020-06-19T14:22:53Z",
            "message": {
                "message": "You have exhausted this month's submission limit!",
                "remaining_time": "1481076.929224"
            }
        }
    ]
}
```
**Screenshots:**
- Remaining Submission Table
screenshot from 2019-02-12 02-03-04
screenshot from 2019-02-12 02-03-46
- Submission Instructions
screenshot from 2019-02-12 12-50-24
screenshot from 2019-02-12 12-50-31

### CODE REVIEW

**Comment by vkartik97** [comment]
@RishabhJain2018 @deshraj Please review

**Comment by vkartik97** [comment]
@RishabhJain2018 @deshraj Please review!

File: apps/jobs/views.py [hunk 1]
```diff
@@ -393,112 +401,47 @@ def leaderboard(request, challenge_phase_split_id):
...
+            request.user, challenge_pk)
+    phases_data['participant_team_id'] = get_participant_team_id_of_user_for_a_challenge(
+            request.user, challenge_pk)
+    phase_data_list = list()
+    for phase in challenge_phase:
+        remaining_submission_message, response_status = get_remaining_submission_for_phase(request.user,
+                                                                                           phase.id,
+                                                                                           challenge_pk)
+        if response_status != status.HTTP_200_OK:
+            return Response(remaining_submission_message, status=response_status)
```
Comment (RishabhJain2018) [inline apps/jobs/views.py]: Why do we need this condition? Am I missing something here?

File: apps/participants/utils.py [hunk 2]
```diff
@@ -33,6 +33,15 @@ def get_participant_team_id_of_user_for_a_challenge(user, challenge_id):
     return None
 
 
+def get_participant_team_name_of_user_for_a_challenge(user, challenge_id):
+    """Returns the participant team object for a particular user for a particular challenge"""
+    participant_team_id = get_participant_team_id_of_user_for_a_challenge(user, challenge_id)
+    if participant_team_id:
+        return ParticipantTeam.objects.get(pk=participant_team_id).team_name
+    else:
```
Comment (RishabhJain2018) [inline apps/participants/utils.py]: Please remove this condition and simply use `return`

File: apps/participants/utils.py [hunk 3]
```diff
@@ -33,6 +33,15 @@ def get_participant_team_id_of_user_for_a_challenge(user, challenge_id):
...
+        None
```
Comment (RishabhJain2018) [inline apps/participants/utils.py]: Please remove this line.

File: apps/jobs/urls.py [hunk 4]
```diff
@@ -13,6 +13,10 @@
         r'phases/(?P<challenge_phase_pk>[0-9]+)/'
         r'remaining_submissions$',
         views.get_remaining_submissions, name='get_remaining_submissions'),
+    url(r'^(?P<challenge_pk>[0-9]+)/'
+        r'remaining_submissions$',
```
Comment (RishabhJain2018) [inline apps/jobs/urls.py]: Please append a slash at the end of URL.

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -1654,6 +1720,10 @@
             vm.stopFetchingSubmissions();
             vm.stopLeaderboard();
         });
+
+        vm.showConfirmation = function(message){
+            $rootScope.notify("success", message);
+        };
     }
 
 })();
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: Please add a new line here.

File: tests/unit/jobs/test_urls.py [hunk 6]
```diff
@@ -166,3 +166,10 @@ def test_get_submission_by_pk(self):
                          '/api/jobs/submission/{}'.format(self.submission.pk))
         resolver = resolve(self.url)
         self.assertEqual(resolver.view_name, 'jobs:get_submission_by_pk')
+
+    def test_get_remaining_submissons_for_all_phases_url(self):
+        self.url = reverse_lazy('jobs:get_remaining_submissions_for_all_phases',
+                                kwargs={'challenge_pk': self.challenge.pk})
+        self.assertEqual(self.url, '/api/jobs/{0}/remaining_submissions'.format(self.challenge.pk))
```
Comment (RishabhJain2018) [inline tests/unit/jobs/test_urls.py]: Please append slash here in the URL.

File: tests/unit/jobs/test_views.py [hunk 7]
```diff
@@ -366,6 +366,21 @@ def test_challenge_submission_when_maximum_limit_exceeded(self):
...
+                                    kwargs={'challenge_id': self.challenge.pk,
+                                            'challenge_phase_id': self.challenge_phase.pk})
+
+            self.challenge.participant_teams.add(self.participant_team)
+            self.challenge.is_docker_based = True
+            self.challenge.save()
+
+            response = self.client.post(self.url, {
+                                        'status': 'submitting', 'input_file': self.input_file}, format="multipart")
+            self.challenge.is_docker_based = False
```
Comment (RishabhJain2018) [inline tests/unit/jobs/test_views.py]: Why do we need this?

File: frontend/src/views/web/challenge/submission.html [hunk 8]
```diff
@@ -20,18 +20,93 @@
...
                 </div>
-                <div class="ev-md-container ev-panel-title"><strong>Please fill the following fields to submit</strong></div>
-                <div class="row">
+            </div>
+            <div class="ev-md-container ev-card-panel ev-z-depth-5 ev-challenge-view eval-card" ng-if="challenge.page.is_docker_based">
+                <span class="fs-18">
+                    <strong>Submission Limits</strong>
+                </span><br>
+                <span class="fs-14">
+                    <strong>Team Name:</strong> {{challenge.phaseRemainingSubmissions.participant_team}}
```
Comment (RishabhJain2018) [inline frontend/src/views/web/challenge/submission.html]: Please change `Team Name:` to `Participant Team`:

File: frontend/src/views/web/challenge/submission.html [hunk 9]
```diff
@@ -20,18 +20,93 @@
...
+                    <strong>Team ID:</strong> {{challenge.phaseRemainingSubmissions.participant_team_id}}
+                </span>
+                <table class="highlight centered">
+                    <thead>
+                        <tr>
+                            <th>Phase</th>
+                            <th>Start Date</th>
+                            <th>End Date</th>
+                            <th>Remaining Submissions</th>
+                        </tr>
```
Comment (RishabhJain2018) [inline frontend/src/views/web/challenge/submission.html]: We also need to add one more column here saying `Command to submit`, which will display the command to submit to that particular phase along with a copy button like for phase with ID 1 `evalai push <image:tag> -p 1`

File: frontend/src/views/web/challenge/submission.html [hunk 10]
```diff
@@ -20,18 +20,93 @@
...
+                    <strong>Team Name:</strong> {{challenge.phaseRemainingSubmissions.participant_team}}
+                </span><br>
+                <span class="fs-14">
+                    <strong>Team ID:</strong> {{challenge.phaseRemainingSubmissions.participant_team_id}}
+                </span>
+                <table class="highlight centered">
+                    <thead>
+                        <tr>
+                            <th>Phase</th>
+                            <th>Start Date</th>
```
Comment (RishabhJain2018) [inline frontend/src/views/web/challenge/submission.html]: Let's merge start date and end date in a single column. Something like: 
screenshot 2019-02-16 at 8 11 38 pm

**Comment by RishabhJain2018** [comment]
@vkartik97 Please do let me know once you have incorporated all the changes.

File: apps/participants/utils.py (hunk 2 above)
Comment (vkartik97) [reply apps/participants/utils.py]: @RishabhJain2018 
As we are getting `participant_team_id` from `get_participant_team_id_of_user_for_a_challenge`, there is a case where we might get where `participant_team_id = None`.

File: tests/unit/jobs/test_views.py (hunk 7 above)
Comment (vkartik97) [reply tests/unit/jobs/test_views.py]: It's not required. Removed it

File: apps/jobs/views.py (hunk 1 above)
Comment (vkartik97) [reply apps/jobs/views.py]: This is there to handle 403 from [`Conditional check for the existence of participant team of the user`](https://github.com/Cloud-CV/EvalAI/blob/c237a07fda872cff5b79975898544ca6250dbae9/apps/jobs/utils.py#L27) or some other checks added later.

File: frontend/src/views/web/challenge/submission.html (hunk 8 above)
[the remark this answers is not in the document]
Comment (vkartik97) [reply frontend/src/views/web/challenge/submission.html]: `Team Name` and `{{challenge.phaseRemainingSubmissions.participant_team}}` have same font size(14px), checked on chrome

**Comment by vkartik97** [comment]
@RishabhJain2018 Please Review!

**Comment by RishabhJain2018** [comment]
Closing this since #2161 is opened for it.

**Pull request closed without merging**